switchio/loop.py

```diff
@@ -262,29 +340,101 @@ def start(self):
         else:
             self.log.error("Unknown event '{}'".format(evname))
 
+    def get_id(self, e, default=None):
+        """Acquire the client/consumer (app) id for event :var:`e`
+        """
+        for var in self.app_id_headers:
+            ident = e.get(var)
+            if ident:
+                self.log.debug(
+                    "app id lookup using '{}' successfully returned '{}'"
+                    .format(var, ident)
+                )
+                return ident
+        return default
+
+    def waitfor(self, sess, varname, timeout=None):
+        '''Wait on a boolen variable `varname` to be set to true for
+        session `sess` as read from `sess.vars['varname']`.
+        This call blocks until the attr is set to `True` most usually
+        by a callback.
+
+        WARNING
+        -------
+        Do not call this from the event loop thread!
+        '''
+        # retrieve cached event/blocker if possible
+        event = mp.Event() if not self._blockers else self._blockers.pop()
+        waiters = self._sess2waiters.setdefault(sess, {})  # sess -> {vars: ..}
```

Comment:
collections.defaultdict might be a good fit here.

Reply:
This is going to get dropped entirely and replaced with something similar like `Session.recv()` (`Session.waitfor(var='blah')` maybe?) which returns a future that's used for inter-coroutine comms (ICC?) heh?

Reply:
Yeah gonna defer this until solving #12